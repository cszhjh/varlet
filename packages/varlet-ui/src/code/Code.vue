<template>
  <div :class="n()">
    <div :class="n('content')" v-html="highlightedCode"></div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref, watch } from 'vue'
import { createNamespace } from '../utils/components'
import { props } from './props'
import { injectHighlighterProvider } from '../highlighter-provider/provide'
import { isFunction } from '@varlet/shared'

const { name, n } = createNamespace('code')

export default defineComponent({
  name,
  props,
  setup(props) {
    const { highlighter, theme } = injectHighlighterProvider()

    const highlightedCode = ref<string | undefined>(props.code)
    const getTheme = computed(() => props.theme || theme)

    isFunction(highlighter?.codeToHtml) &&
      watch(
        () => [props.code, props.language, getTheme.value],
        async ([code, lang, theme]) => {
          highlightedCode.value = await highlighter.codeToHtml(code || '', { lang, theme })
        },
        { immediate: true }
      )

    return {
      highlightedCode,
      n,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import './code';
</style>
