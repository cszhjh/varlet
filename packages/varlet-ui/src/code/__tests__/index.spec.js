import Code from '..'
import VarCode from '../Code'
import VarHighlighterProvider from '../../highlighter-provider'
import { mount } from '@vue/test-utils'
import { createApp, h } from 'vue'
import { delay } from '../../utils/test'
import { expect, describe, test } from 'vitest'
import { codeToHtml } from 'shiki'

const code = `function twoSum(nums, target) {
  const map = new Map();
  for (let i = 0; i < nums.length; i++) {
      const num = nums[i];
      const theOther = target - num;
      if (map.has(theOther)) {
          return [map.get(theOther), i];
      }
      map.set(num, i);
  }
};`

test('test code use', () => {
  const app = createApp({}).use(Code)
  expect(app.component(Code.name)).toBeTruthy()
})

describe('test code component props', () => {
  test('test code content', async () => {
    const wrapper = mount(VarHighlighterProvider, {
      props: {
        highlighter: {
          codeToHtml,
        },
      },
      slots: {
        default: () =>
          h(VarCode, {
            props: {
              code,
              language: 'javascript',
              theme: 'monokai',
            },
          }),
      },
    })

    await delay(300)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  test('test code lang', async () => {
    const wrapper = mount(VarHighlighterProvider, {
      props: {
        highlighter: {
          codeToHtml,
        },
      },
      slots: {
        default: () =>
          h(VarCode, {
            props: {
              content: code,
              lang: 'javascript',
            },
          }),
      },
    })

    await delay(300)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  test('test code theme', async () => {
    const wrapper = mount(VarHighlighterProvider, {
      props: {
        highlighter: {
          codeToHtml,
        },
      },
      slots: {
        default: () =>
          h(VarCode, {
            props: {
              code,
              language: 'javascript',
              theme: 'material-theme',
            },
          }),
      },
    })

    await delay(300)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })
})
