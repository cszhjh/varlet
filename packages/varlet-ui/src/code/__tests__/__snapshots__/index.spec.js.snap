// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test code component props > test code content 1`] = `
"<div class="var-highlighter-provider">
  <div class="var-code" props="[object Object]">
    <div class="var-code__content"><pre class="shiki nord" style="background-color:#2e3440ff;color:#d8dee9ff" tabindex="0"><code><span class="line"><span></span></span></code></pre>
    </div>
  </div>
</div>"
`;

exports[`test code component props > test code lang 1`] = `
"<div class="var-highlighter-provider">
  <div class="var-code" props="[object Object]">
    <div class="var-code__content"><pre class="shiki nord" style="background-color:#2e3440ff;color:#d8dee9ff" tabindex="0"><code><span class="line"><span></span></span></code></pre>
    </div>
  </div>
</div>"
`;

exports[`test code component props > test code theme 1`] = `
"<div class="var-highlighter-provider">
  <div class="var-code" props="[object Object]">
    <div class="var-code__content"><pre class="shiki nord" style="background-color:#2e3440ff;color:#d8dee9ff" tabindex="0"><code><span class="line"><span></span></span></code></pre>
    </div>
  </div>
</div>"
`;
